app/models/piece.rb

```diff
@@ -101,4 +121,44 @@ def enemy_at(check_row, check_col)
     enemy = game.piece_at(check_row, check_col)
     enemy if enemy && enemy.is_black != is_black
   end
+
+  def get_path(to_row, to_col)
+    return nil if %w(Rook Bishop Queen).exclude?(type)
+    path = []
+    current_row = row
+    current_col = col
+    until current_row == to_row && current_col == to_col
+      current_row += to_row <=> current_row
+      current_col += to_col <=> current_col
+      path << [current_row, current_col]
+    end
+    path[0..-2]
+  end
+
+  protected
+
+  def valid_move?(to_row, to_col)
+    return false if move_nil?(to_row, to_col)
+    return false if move_out_of_bounds?(to_row, to_col)
+    return false if move_destination_ally?(to_row, to_col)
+    return false unless move_legal?(to_row, to_col)
+    return false if move_obstructed?(to_row, to_col)
+    true
+  end
+
+  def self_check?(to_row, to_col)
+    return false unless move_to!(to_row, to_col)
```

Comment:
This means that for each call to `self_check?`, there will be several database queries (update piece, create move, update piece, destroy move, possibly update captured piece) so there's a lot of DB traffic which will affect performance and will have a lot of surface area for something to crash in the middle and leave the DB in an inconsistent state.

However, due to our time constraints (hey, it simulates a real-world project!) I think if this is refactored, it should be done in a later PR as that would likely be time-intensive and involve some kind of temporary in-memory representation of the board which would affect pretty much all of the move logic.

So, just wanted to note this so that it's in the GitHub log, but I don't think it needs to be resolved in this PR.